pkg/topology/kademlia/kademlia.go

```diff
@@ -231,43 +231,35 @@ func (k *Kad) connectBalanced(wg *sync.WaitGroup, peerConnChan chan<- *peerConnI
 		}
 
 		binPeers := k.knownPeers.BinPeers(uint8(i))
+		binConnectedPeers := k.connectedPeers.BinPeers(uint8(i))
 
 		for j := range k.commonBinPrefixes[i] {
 			pseudoAddr := k.commonBinPrefixes[i][j]
 
-			closestConnectedPeer, err := closestPeer(k.connectedPeers, pseudoAddr, noopSanctionedPeerFn)
+			// Connect to closest known peer which we haven't tried connecting to recently.
+
+			_, err := nClosePeerInSlice(binConnectedPeers, pseudoAddr, noopSanctionedPeerFn, uint8(i+k.bitSuffixLength+1))
 			if err != nil {
-				if errors.Is(err, topology.ErrNotFound) {
-					break
+				if !errors.Is(err, topology.ErrNotFound) {
+					break // not possible
```

Comment:
technically if there is an error here that's not topology.ErrNotFound, we should return, instead of break.